Allow JT specification and prompting for project branch

AlanCoding opened the pull request:
##### SUMMARY
Implements the core of https://github.com/ansible/awx/issues/282

Decisions that are non-obvious from the prior discussion:

 - using the default branch with `allow_override` being True will behave the same as other projects (offline runs are possible). I still would argue that this isn't consistent, and will be a surprise to the user, but any decision here would have problems with communication.
 - prompting in workflows is allowed in the node prompts, same with schedules
 - playbook list is produced from project's default branch (for use as suggestions)

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API


##### ADDITIONAL INFORMATION
We hope to handle in another PR:

 - organization as field on JT
 - cloning from a PR ref (may be done in Ansible core)

AlanCoding committed:
Allow JTs to specify and prompt for SCM branch

Copy project folder each job run
  change cwd to private_data_dir, from proj
  do not add cwd to show_paths if it is
  a subdirectory of private_data_dir, which
  is already shown

Pass the job private_data_dir to the local
  project sync, and also add that directory
  to the project sync show paths

Add GitPython dep and use for job sync logic
  use this to manage shallow clone from desired
  commit, and to map branch to commit,
  and to assess necessity of project sync

Start on some validation change, but not all
  allow arbitrary playbooks with custom branch

Commit:
assorted UI work to support the new branch field
update project to have allow branch override checkbox
add new text input for branch field
adjust show/hide for branch and playbook jt fields
make playbook field allowed to add a new option not in the dropdown
update job results ui to show branch
update prompting to support new branch field

AlanCoding committed:
Add scm_revision to project updates and cleanup

Add validation around prompted scm_branch requiring
  project allow_override field to be true

Updated related process isolation docs

Fix invalid comarision in serializer

from PR review, clarify pre-check logging, minor docs additions

matburt commented:
Can you explain this more? 

> using the default branch with allow_override being True will behave the same as other projects (offline runs are possible). I still would argue that this isn't consistent, and will be a surprise to the user, but any decision here would have problems with communication.

matburt commented on awx/api/serializers.py:
```diff
@@ -1388,6 +1388,11 @@ def get_field_from_model_or_attrs(fd):
+        if 'allow_override' in attrs and self.instance:
+            if attrs['allow_override'] != self.instance:
+                raise serializers.ValidationError({
+                    'allow_override': _('Branch override behavior of a project cannot be changed after creation.')})
```
Why? Even if we know why we need to be able to tell people why.

kdelee replied:
1) Not sure I like this
2) dont' show me button if I can't do it
3) This is not coming up through UI, getting mysterious 400 instead:
![Screenshot from 2019-07-24 14-05-29](https://user-images.githubusercontent.com/29239279/61817522-c3466a80-ae1c-11e9-92c4-0c569ccacd55.png)

AlanCoding replied:
Let me recap the reason that this was added, say someone does this:

 - make project with `allow_override` being true
 - makes JT with a non-blank `scm_branch`
 - changes project `allow_override` field back to False

Now the user has a JT that specifies a branch that we cannot apply.

Providing an outline of this case is far too much wording to put in an error message, but I agree the UI should at least show the error message from the API. Or hide it when in edit mode.

Also, this is more complicated than the other "prompts" related rejections of fields. If a JT's field is rejected because the project is configured to not allow it, we currently have no mechanism or precedent for dealing with that. We have no way of letting the user know that's broken without a lot of novel UI work.

AlanCoding replied:
After more discussions, I agree the API side of this can be made smoother for the user interaction.

It would be fairly easy to combine a check with this, something like `JobTemplate.objects.filter(project=obj).exclude(scm_branch="").exists()`, and then if both this condition is satisfied, and the `allow_override` value changes from its prior value, then we could give a more specific error message:

> Branch override behavior of this project is currently relied on by one or more existing job templates.

Then if there are no such JTs, we can just allow the PATCH/PUT.

dsesami replied:
+1 to @kdelee's view of things and your solution seems to be a step in the right direction @AlanCoding , it seems allowing the change makes more sense for the end user -- it would also make it easier if the dependent templates could be listed, though I'm unsure of performance impact.

AlanCoding replied:
> it would also make it easier if the dependent templates could be listed

That's possible, but depends on how.

If all you're suggesting is a list of names or primary keys in the error message, then sure. That is super easy, and IMO no performance impact that matters. But this will not be linked.

Giving primary keys would be the most concise. If the message is too long, it will overflow the popup box, because error messages are intended to be concise. But if I give primary keys, I worry that someone will say that's a bad UX. But I think handling any special UI cases here is going way overboard for a corner case.

dsesami replied:
@AlanCoding I see the issue. How about something like a warning that existing job templates use branch override behavior, and offer a confirm button that would revert them if the person really doesn't care? If that's a lot to handle and offers chaos, then just the warning would be good though.

matburt commented on awx/api/serializers.py:
```diff
@@ -2701,7 +2706,7 @@ def get_summary_fields(self, obj):
 
     class Meta:
-        fields = ('*', 'job_type', 'inventory', 'project', 'playbook',
+        fields = ('*', 'job_type', 'inventory', 'project', 'playbook', 'scm_branch',
```
This needs to be consistent... it's `scm_revision` elsewhere, which is more appropriate... it's more than just `branch`.

AlanCoding replied:
`JobSerializer` has both fields `scm_branch` and `scm_revision`. The field name `scm_branch` was borrowed from the `Project` model.

The field `scm_revision` is the SHA1 hash or technical specifier of the commit that a job is ran against (like "r49" for subversion). This serves the purpose of audibility in AWX / Tower so that someone can look back in time and be sure they can recover the exact version of the playbook that was ran.

`scm_branch` is given by a user at some point. It can be several things, and the Ansible git module goes through fallback logic where it checks if it is (1) a branch (2) a tag (3) assume it's a commit, and possibly error. The fallback logic for evaluating an scm_branch here is modeled after their process.

matburt commented on awx/main/models/jobs.py:
```diff
@@ -396,7 +418,7 @@ def _accept_or_ignore_job_kwargs(self, **kwargs):
                 # behavior of putting them in ignored_fields and launching anyway
                 if 'prompts' not in exclude_errors:
-                    errors_dict[field_name] = _('Field is not configured to prompt on launch.').format(field_name=field_name)
+                    errors_dict[field_name] = _('Field is not configured to prompt on launch.')
```
🤣

ryanpetrello replied:
Whoops.

matburt commented on awx/main/tasks.py:
```diff
@@ -41,6 +43,10 @@
 from crum import impersonate
 
+# GitPython
+import git
```
I think I may need some help understanding what is going on in this file.... is it just Runner usage cleanup?

matburt commented on awx/main/tasks.py:
```diff
@@ -1563,46 +1569,131 @@ def should_use_proot(self, job):
+                except (ValueError, BadGitName):
+                    pass
+            else:
+                if git_repo.head.commit.hexsha == job.project.scm_revision:
```
Is this to cover the scenario where they gave a specific revision? Rather than a branch or a tag?

AlanCoding commented:
Using example:

https://github.com/ansible/ansible/branches

Steps:

 - I create a project from the Ansible core repo with
    - `allow_override` is True
    - default `scm_branch` is `devel`
 - Right after creating it
    - I run a job without overriding the branch
    - I run a job that runs with `stable-2.8` branch
 - I return a week later and relaunch both of those jobs
   - `devel` **does not pull a newer revision**
   - `stable-2.8` does pull the latest revision from github and runs that

This is because we run against the revision of the last "check"-type project update, which has its revision stored. Prompted, non-default branches, do not have their revision saved.

We can change that, but a lot of question need to be answered. This behavior both was the least invasive, and closest to the requirements discussed in planning. I wanted to give pushback on this, but I'm happy with it now. While it's surprising that the default branch behaves differently from the non-default branch, it would also be surprising that different values of `allow_override` would behave differently. So unless we save the revision of all branches (technically problematic in some cases, impossible in others), the current implementation seems like the best option.

I might take some of this and put it in the docs folder, that had slipped my mind so-far.

matburt commented on awx/main/tasks.py:
```diff
@@ -1896,25 +2006,46 @@ def acquire_lock(self, instance, blocking=True):
+                project_path = p.get_project_path(check_if_exists=False)
+                git_repo = git.Repo(project_path)
+                try:
+                    commit = git_repo.commit(instance.scm_branch)
```
I assume this works on all ref types like tags and branches?

matburt reviewed (commented):
There's a lot going on in `tasks.py` here... I might want you to annotate what's going on and what you are working towards here, especially since it's going to affect project syncing overall.

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -2748,16 +2753,28 @@ def to_representation(self, obj):
     def validate(self, attrs):
         if 'project' in self.fields and 'playbook' in self.fields:
-            project = attrs.get('project', self.instance and self.instance.project or None)
+            project = attrs.get('project', self.instance.project if self.instance else None)
```
Yes please.

ryanpetrello commented on awx/main/migrations/0082_v360_job_branch_overrirde.py:
```diff
@@ -0,0 +1,46 @@
+        migrations.AlterField(
+            model_name='project',
+            name='scm_update_cache_timeout',
+            field=models.PositiveIntegerField(blank=True, default=0, help_text='The number of seconds after the last project update ran that a new project update will be launched as a job dependency.'),
```
This only changed because the `help_text` got an extra space in it?

AlanCoding replied:
Yes, this is correct, two words ran together. I figured this is a good time to fix it because I was already migrating fields on the model.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1563,46 +1569,131 @@ def should_use_proot(self, job):
+                    logger.info('Skipping project sync for {} because commit is locally available'.format(job.log_format))
+                    needs_sync = False  # requested commit is already locally available
+                except (ValueError, BadGitName):
+                    pass
```
Should we put a `logger.error()` here so people can see in the logs that they put in an invalid ref?

Commit:
fix ui conditional for adding fields to jt edit save payload

jbradberry commented on awx/main/tasks.py:
```diff
@@ -1563,46 +1569,118 @@ def should_use_proot(self, job):
+            # manual projects are not synced, user has responsibility for that
+            needs_sync = False
+        elif not os.path.exists(project_path):
+            logger.debug('Performing fresh clone of {} on this instance.'.format(job.project))
```
These added logging statements here all seem like they should probably be info-level, not debug.

AlanCoding replied:
https://github.com/ansible/awx/pull/3882/files#diff-8f5e2fe91e83c971fc84cb35476362db

A lot of the INFO-level logs got axed en masse (almost all, really), as per the reasons there. I raised the level of some select log statements in this PR. For "Skipping project sync", I can buy INFO, because it explains the lack of a unified job. But for statements like this one, they fit in the category of normal cluster operation and that seems to fit in the DEBUG stream precedent.

ryanpetrello replied:
I agree `DEBUG` is more appropriate here - this is just informational regarding work we're doing, and it's not an abnormal thing to have happen.  Raising the log level will probably make the logs pretty noisy.

AlanCoding committed:
Add scm_refspec field

Update migration syntax to Django 2

fix status bug where canceled switched to error

AlanCoding commented:
recheck

AlanCoding commented:
Hit expected conflicts with merge of @chrismeyersfsu https://github.com/ansible/awx/pull/4301, pushing resolution.

chrismeyersfsu committed:
use ansible nightly

* ansible:devel now has ansible-galaxy collection support

chrismeyersfsu committed:
collections/requirements.yml support

* just like we support ansible-galaxy role install, support
ansible-galaxy collection install

AlanCoding committed:
Fix bugs with discovery of collection requirements

Addresses some cases where
  collection requirements do not exist
  collection requirements cannot be evaluated

Consolidate logic for roles and collection installs

chrismeyersfsu committed:
force proj sync when collections/requirements.yml

* Similar to roles/requirements.yml sync optimization logic.

AlanCoding committed:
JT-branch docs and code cleanup

bump migration

fine tune validation of project allow_override
  return highly custom error message

Restore branch after syncs to address bugs
  encountered after changing scm_refspec

remove unused code to determine scm_revision

Check Ansible version before project update and
  do not install collections if Ansible version too old

Add docs related to project branch override
  New file specific to branch override and refspec
Complete docs on collections to reflect current
  implementation and give a folder tree example
Update clustering docs related to project syncs

Fix bug where git depth was ignored during the
  local clone from project folder to run folder

Fix bug where submodules were not copied

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -1311,6 +1311,8 @@ def validate(self, attrs):
         if 'local_path' in attrs and attrs['local_path'] not in valid_local_paths:
             errors['local_path'] = _('This path is already being used by another manual project.')
+        if attrs.get('scm_refspec') and scm_type != 'git':
+            errors['scm_refspec'] = _('SCM refspec can only be used with git projects.')
```
@wenottingham is this a requirement you're comfortable with in the final version of this?  I know we were all quite skeptical about SVN, but there's still hg to think about.

Just to be clear and set expectations, this current version of this PR *only* supports git (not hg).

wenottingham replied:
Yes, I'm OK with that.

ryanpetrello commented on awx/main/tests/functional/api/test_job_template.py:
```diff
@@ -505,3 +505,37 @@ def test_callback_disallowed_null_inventory(project):
+        user=admin_user,
+        expect=400
+    )
+    assert 'Project does not allow overriding branch' in str(r.data['ask_scm_branch_on_launch'])
```
👍

ryanpetrello commented on awx/main/tests/functional/api/test_project.py:
```diff
@@ -44,3 +45,52 @@ def test_project_unset_custom_virtualenv(get, patch, project, admin, value):
+        user=admin_user,
+        expect=200
+    )
+    assert Project.objects.get(pk=r1.data['id']).allow_override is False
```
👍

ryanpetrello commented on awx/playbooks/project_update.yml:
```diff
@@ -148,18 +126,29 @@
+      delegate_to: localhost
+
+    - block:
+      # TODO: don't run this if ansible verison doens't support collections
```
Typo: version

ryanpetrello replied:
Actually, can't we just lose this TODO now?

https://github.com/ansible/awx/pull/4265/files#diff-9d4ea1dd908b35fb92eaede4bd10bb46R1893

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -692,9 +698,12 @@ class BaseTask(object):
     proot_show_paths = []
 
+    def __init__(self, *args, **kwargs):
+        super(BaseTask, self).__init__(*args, **kwargs)
```
Do you need this line?  The parent class is just `object`.

AlanCoding replied:
...since this is `BaseTask`, not some derived class, I agree strongly. Will push after re-runs of tests.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -858,7 +869,10 @@ def build_params_process_isolation(self, instance, private_data_dir, cwd):
+            local_paths = [private_data_dir]
+            if cwd != private_data_dir and Path(private_data_dir) not in Path(cwd).parents:
+                local_paths.append(cwd)
+            show_paths = self.proot_show_paths + local_paths + \
```
👍

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1575,46 +1590,119 @@ def should_use_proot(self, job):
+            source_branch = git_repo.create_head(tmp_branch_name, job.scm_revision)
+            git_repo.clone(runner_project_folder, branch=source_branch, depth=1, single_branch=True)
+            # force option is necessary because remote refs are not counted, although no information is lost
+            git_repo.delete_head(tmp_branch_name, force=True)
```
This is sneaky and I like it 😄

AlanCoding replied:
FYI, in more recent git versions you can clone directly from a SHA-1 hash. If we could count on that functionality, this sneaky part would not really be necessary. We would just know what rev the job is going to run, and tell it "put that here". We could change to the new style in the future after supported platforms have upgraded git, and make this a little cleaner.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1575,46 +1590,119 @@ def should_use_proot(self, job):
+            galaxy_collections_req_path = os.path.join(project_path, 'collections', 'requirements.yml')
+            if os.path.exists(galaxy_collections_req_path):
+                logger.debug('Running project sync for {} because of galaxy collections requirements.'.format(job.log_format))
+                needs_sync = True
```
This `needs_sync` logic all makes sense to me.

ryanpetrello commented on awx/main/tasks.py:
```diff
@@ -1651,7 +1739,24 @@ class RunProjectUpdate(BaseTask):
+    def event_handler(self, event_data):
+        super(RunProjectUpdate, self).event_handler(event_data)
+        returned_data = event_data.get('event_data', {})
+        if returned_data.get('task_action', '') == 'set_fact':
```
Normally I get nervous about custom event handling for performance reasons, but if this sort of thing is restricted to project update events (it seems to be), I think we're okay.

ryanpetrello replied:
Also, nice defensive programming here w/ all the `.get()`s

ryanpetrello replied:
All of this said, I _think_ that runner actually exposes fact cache data via another mechanism that might be easier than what you're doing here.

If you set `ANSIBLE_CACHE_PLUGIN = json` on the project update run, you should be able to do this sort of thing:

Example: https://github.com/ansible/awx/commit/99478f5d255986cba796b78551a4e2eaf5ef5f17#diff-95c3599b75db21867b3991cee7b89575R272

Maybe that's more work than this approach, though.

AlanCoding replied:
Good points, it definitely crossed my mind. To start from the requirements - I want a simple-as-possible programatic pure-python data pipeline for values to come from my playbook, through runner, to the code that is calling runner. I landed on the event callback, but we can totally go a different direction if a better option presents itself. I don't remember exactly, but I was having some headaches with the bubblewrap / error handling of the old writing to file method.

ryanpetrello replied:
Yep, I'm fine with what's in this PR as-is.

dsesami commented:
cc @jlmitch5 -- currently, when setting prompt on launch for a branch override in a JT, and then using that JT in a workflow, the workflow never prompts me for the branch value. Something to look at during the next round of implementation.

ryanpetrello reviewed (approved):
This is fancy.

I was able to use this to launch a JT, specifying the branch at launch time:

```shell
$ awx projects get 4 -f human --filter scm_url
=========================================
scm_url
=========================================
https://github.com/ansible/test-playbooks
=========================================
$ awx projects modify 4 --allow_override 1
$ awx job_templates modify 5 --ask_scm_branch_on_launch 1 --playbook assert_on_this_branch.yml
$ awx job_templates launch 5 --scm_branch empty_branch --monitor -f human
```

https://github.com/ansible/test-playbooks/blob/empty_branch/assert_on_this_branch.yml

<img width="720" alt="image" src="https://user-images.githubusercontent.com/214912/62224875-5df6fa00-b385-11e9-8766-c912108f3b64.png">

If you specify an invalid branch, the project update fails in a  _somewhat_ unclear way

```shell
$ awx job_templates launch 5 --scm_branch invalid_branch
```

<img width="714" alt="image" src="https://user-images.githubusercontent.com/214912/62225370-4e2be580-b386-11e9-9e21-c1dca2410e20.png">

<img width="396" alt="image" src="https://user-images.githubusercontent.com/214912/62225386-53893000-b386-11e9-833a-109839ad541c.png">

Since it's an implicit project update failure, you've actually got to go dig in the database:

<img width="720" alt="image" src="https://user-images.githubusercontent.com/214912/62225448-73b8ef00-b386-11e9-8dc7-3d25d7e9fd83.png">

Maybe we could come up with a better way to surface this information?

ryanpetrello commented:
Actually @AlanCoding, @jakemcdermott pointed out to me that _this_ exists:

![image](https://user-images.githubusercontent.com/214912/62232662-c4cfdf80-b394-11e9-975a-ed934f89a30a.png)

...so maybe that's good enough?

![image](https://user-images.githubusercontent.com/214912/62232692-d0bba180-b394-11e9-9fc4-f6216283a8ab.png)

Commit:
working commit

AlanCoding commented:
Let me give an update on a particular use case that I dived into, and have concluded _cannot_ be done, and reasons why.

A project admin _can not_ use the `scm_refspec` field to _restrict_ what branches from that project can be used. In other words, if you set allow_override=True but provide a refspec that only fetches 1 branch, a job template or job can still provide any normal branches or tags.

This behavior goes back to the Ansible git module, which goes back to Ansible. The key as to why we can or cannot comes down to a command like this:

```
git clone -c remote.origin.fetch=+refs/heads/master:refs/remotes/origin/master https://github.com/ansible/test-playbooks.git
```

What would you expect this to do? A single branch clone, or should it fetch all branches? It fetches all branches because it applies the config:

```
[remote "origin"]
	fetch = +refs/heads/master:refs/remotes/origin/master
	url = https://github.com/ansible/test-playbooks.git
	fetch = +refs/heads/*:refs/remotes/origin/*
```

So by normal means, we can only add a refspec to fetch, we cannot _replace_ the default.

Now I still see a valid argument for the Ansible git module to either add the `single_branch` option, or to automatically clone with that argument when a `refspec` is provided. This is untested, but even if it works it'll be a watered down request.

I have attempted to make the docs here sufficiently general so that I do not imply that a `refspec` may restrict the available set of branches.

Commit:
more ui work for branch and refspec on project/jt
- add refspec field to project
- update refspec and branch help text on project form
- add refspec field to job detail
- adjust form gen and ProcessErrors to show api errors for checkbox_groups correctly
- consolidate showPromptButton conditionals and fix the add/edit workflow node one for showing prompt when only branch is promptable

AlanCoding committed:
avoid unnecessary checkout, more docs content

Commit:
fix issue with interior scope declaration eslint error

AlanCoding commented:
recheck

AlanCoding committed:
Add collection setting toggle to UI

Additional API housekeeping, removing unused code

Treat default branch as no branch provided

ryanpetrello commented:
@jlmitch5 Is there any other UI work required?  If not, do we have somebody lined up to review the UI changes?

I'm ready to merge this whenever @tvo318 @squidboylan @kdelee sign off 👍

Commit:
remove inadverdent duplicate CreateSelect2 call from playbook on jt edit form

jakemcdermott commented:
After enabling `allow branch override`, I tried changing the demo project to point to `https://github.com/jlaska/ansible-playbooks` and got this error:
![Screenshot from 2019-08-09 15-08-13](https://user-images.githubusercontent.com/9753817/62803249-ed07be80-bab7-11e9-98cc-ffe635cb041b.png)

Is this expected / already known?

jakemcdermott commented:
Something else I ran into:

- I enabled `allow branch override` on the default project, which is related to the demo job template
- Not making any other changes to any other resource, I attempted to _undo_ my previous action by unchecking the `allow branch override` box and saving. This resulted in an error:

![Screenshot from 2019-08-09 15-13-11](https://user-images.githubusercontent.com/9753817/62803367-3fe17600-bab8-11e9-86f0-2350b6297474.png)

Some comments on this behavior:
- I _haven't_ specified a branch for this project on the related job template. It's completely blank because I haven't changed anything.
- A user should be able to undo an action like this, especially if they haven't changed anything else since performing the action they want to undo.

AlanCoding commented:
@jakemcdermott in your 2nd screenshot, it does give a meaningful message under the field. Don't know about expectations with the pop up.

Also, `https://github.com/jlaska/ansible-playbooks` is a redirect. I don't know if this was pre-existing behavior that these fail.

jakemcdermott commented:
@AlanCoding 

- The message is meaningful but inaccurate - I _haven't_ specified a branch on the related JT. It's blank.
- I considered that the redirect might be a factor, but I see the same problem when using `https://github.com/ansible/test-playbooks`. I don't think it is pre-existing behavior for these to fail as I use them regularly on devel.

AlanCoding commented:
> The message is meaningful but inaccurate - I haven't specified a branch on the related JT. It's blank.

Did you set the JT to prompt for branch on launch? We still have to block the action in this case.

Say we change it to:

> One or more job templates depend on branch override behavior for this project (ids: 38, 4, 99).

Is that okay to be the final wording? @tvo318 does this look good enough to ship to you?

In terms of how to resolve the unwanted pop-up I defer to @jlmitch5

jakemcdermott commented:
scm branch is not set for prompting on the JT - I'll take a moment on monday to try and reproduce this on a fresh build

jakemcdermott commented:
@AlanCoding I was able to reproduce this after doing a fresh rebuild. Below are some more exact steps:

### setup
terminal 1:
```shell
docker rm -f $(docker ps -a -q)
docker rmi -f $(docker images -a -q)

rm -rf awx
git clone https://github.com/ansible/awx

cd awx
git remote add alan git@github.com:AlanCoding/awx.git
git fetch alan
git checkout branch_feature_phase_2
cp awx/settings/local_settings.py.docker_compose awx/settings/local_settings.py
COMPOSE_TAG=branch_feature_phase_2 make docker-compose-build
COMPOSE_TAG=branch_feature_phase_2 make docker-compose
```

terminal 2:
(after waiting for environment to finish spinning up in terminal 1)
```shell
docker exec -i tools_awx_1 sh <<-EOSH
  awx-manage createsuperuser --noinput --username=admin --email=null@ansible.com
  awx-manage update_password --username=admin --password=password
  awx-manage create_preload_data
EOSH

cd awx
make ui-devel
```

###  steps
1. open https://localhost:8043 in a browser
2. sign in as `admin` / `password`
3. navigate to projects page
4. run a project update on the demo project and wait for it to finish
5. open the details form for the project
6.  change the project url to `https://github.com/ansible/test-playbooks`
7.  save
8. check the box for `allow branch override`
9. save
10. uncheck the box for `allow branch override`
11. save :x: 

### expected results
The form saves successfully, indicating that a user can undo their decision to check the box. 

### actual results
API errors (see below). The content or correctness of the error message is inconsequential. A user should always be able to undo their actions in this way.

### additional information
error shown after step 11:
![Screenshot from 2019-08-09 15-13-11](https://user-images.githubusercontent.com/9753817/62803367-3fe17600-bab8-11e9-86f0-2350b6297474.png)

---

 project details after step 11:
![Screenshot from 2019-08-12 10-59-06](https://user-images.githubusercontent.com/9753817/62875160-84575680-bcf0-11e9-9f3b-9f86895ec9da.png)
[api response](https://gist.githubusercontent.com/jakemcdermott/5f865bd03cfa807ddf5b4716bfd95a5d/raw/5e1a391dc06a614cd0e3d5ef8a384a60f8770911/4265_project.json)

---

job template details after step 11:
![Screenshot from 2019-08-12 10-45-55](https://user-images.githubusercontent.com/9753817/62875164-86b9b080-bcf0-11e9-8a8c-3f368928c550.png)
[api response](https://gist.githubusercontent.com/jakemcdermott/802f1a2a0306fcea04aada8d2e4448df/raw/477033b22c789b229b521aaf77b977786e989f89/4265_job_template.json)

AlanCoding committed:
Fix logic for turning off override behavior

jakemcdermott reviewed (approved):
- ui code changes look good to me
- https://github.com/ansible/awx/pull/4265#issuecomment-520462639 should be resolved by https://github.com/ansible/awx/pull/4265/commits/addeff2cfaaea21d3768b6d84e8204e32eb3557e
- double display of errors (both under the field _and_ in an alert) is more than likely a pre-existing issue with our error handling utility so I've made a github issue for it https://github.com/ansible/awx/issues/4464

matburt reviewed (approved):
This lgtm
